File: examples/app-baseline/golang/ids_test.go
package main

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"testing"
)

func TestUUIDFromOrdinalMatchesRustLayout(t *testing.T) {
	value := uuidFromOrdinal(0x10, 0)
	if len(value) != 16 {
		t.Fatalf("len=%d", len(value))
	}
	if value[0] != 0x10 {
		t.Fatalf("byte0=%x", value[0])
	}
	if value[6] != 0x70 {
		t.Fatalf("version nibble=%x", value[6])
	}
	if value[8]&0xc0 != 0x80 {
		t.Fatalf("variant=%x", value[8])
	}
	if got := formatUUID(value); got != "10101010-1010-7010-8000-000000000000" {
		t.Fatalf("format=%s", got)
	}
}

func TestFingerprintIsCanonicalAndLengthPrefixed(t *testing.T) {
	first := safeInputFingerprint([][]byte{[]byte("a/b"), []byte("c")})
	second := safeInputFingerprint([][]byte{[]byte("a"), []byte("b/c")})
	repeated := safeInputFingerprint([][]byte{[]byte("a/b"), []byte("c")})
	if len(first) != 64 {
		t.Fatalf("len=%d", len(first))
	}
	if first != repeated {
		t.Fatal("fingerprint was not stable")
	}
	if first == second {
		t.Fatal("length prefix was ignored")
	}
	if strings.Contains(first, "a/b") {
		t.Fatal("raw field leaked into digest")
	}
}

func TestSmokeAndFullSeedShapes(t *testing.T) {
	smoke := generateSeed(smokeScale())
	if len(smoke.organizations) != 2 {
		t.Fatalf("orgs=%d", len(smoke.organizations))
	}
	if smoke.scale.boardDenseOpen != 0 {
		t.Fatalf("board dense=%d", smoke.scale.boardDenseOpen)
	}
	if boardDenseOpenCount(smoke) >= 50 {
		t.Fatalf("smoke board is unexpectedly dense")
	}
	probes := probesFor(smoke)
	if probes.ticketID == probes.writeTicketID {
		t.Fatal("write ticket collided with probe ticket")
	}
	full := generateSeed(fullScale())
	if boardDenseOpenCount(full) != 600 {
		t.Fatalf("full board=%d", boardDenseOpenCount(full))
	}
	if got := len(tenantProbes(full, 3)); got != 3 {
		t.Fatalf("tenants=%d", got)
	}
}

func TestSchemaContainsSafeAppObligations(t *testing.T) {
	for _, table := range []string{
		"app_permission",
		"app_idempotency",
		"app_audit",
		"app_domain_event",
		"app_outbox_intent",
	} {
		if !strings.Contains(schemaSQL, "CREATE TABLE "+table) {
			t.Fatalf("missing table %s", table)
		}
	}
	want := []string{
		"symbolic_operation_authorization",
		"idempotency_admission_and_equal_input_replay",
		"domain_mutation",
		"audit_and_provenance",
		"domain_event",
		"outbox_intent",
		"one_atomic_transaction",
	}
	if len(obligations) != len(want) {
		t.Fatalf("obligations=%v", obligations)
	}
	for i := range want {
		if obligations[i] != want[i] {
			t.Fatalf("obligations=%v", obligations)
		}
	}
}

func TestGoSchemaTablesMatchRustAdapter(t *testing.T) {
	rust, err := os.ReadFile(filepath.Join("..", "postgres", "src", "lib.rs"))
	if err != nil {
		t.Fatal(err)
	}
	tableRe := regexp.MustCompile(`CREATE TABLE (\w+)`)
	rustTables := map[string]struct{}{}
	for _, match := range tableRe.FindAllStringSubmatch(string(rust), -1) {
		rustTables[match[1]] = struct{}{}
	}
	goTables := map[string]struct{}{}
	for _, match := range tableRe.FindAllStringSubmatch(schemaSQL, -1) {
		goTables[match[1]] = struct{}{}
	}
	if len(rustTables) != len(goTables) {
		t.Fatalf("rust=%v go=%v", rustTables, goTables)
	}
	for table := range rustTables {
		if _, ok := goTables[table]; !ok {
			t.Fatalf("missing table %s", table)
		}
	}
}

func TestInteractiveWeightsAreMostlyReads(t *testing.T) {
	total := 0
	writes := 0
	hasSwap := false
	for _, item := range interactiveWeights {
		total += item.weight
		switch item.name {
		case "create_comment", "close_ticket_with_comment", "open_ticket_with_labels":
			writes += item.weight
		case "swap_member_roles":
			hasSwap = true
		}
	}
	if total-writes <= writes {
		t.Fatal("interactive mix is not read-heavy")
	}
	if hasSwap {
		t.Fatal("swap_member_roles must stay out of the interactive mix")
	}
}

func TestRiffdbDriverContainsNoGoSafetyImplementation(t *testing.T) {
	source, err := os.ReadFile("riffdb.go")
	if err != nil {
		t.Fatal(err)
	}
	text := string(source)
	for _, forbidden := range []string{
		"app_permission",
		"app_idempotency",
		"app_audit",
		"app_domain_event",
		"app_outbox_intent",
		"permissionSQL",
		"idempotencyLockSQL",
		"insertAuditSQL",
		"insertEventSQL",
		"insertOutboxSQL",
		"safeAdmit",
		"safeComplete",
		"safeInputFingerprint",
	} {
		if strings.Contains(text, forbidden) {
			t.Fatalf("riffdb.go contains %s", forbidden)
		}
	}
	if !strings.Contains(text, "riffdb.dev/ticketdesk") {
		t.Fatal("generated TicketDesk client is unused")
	}
}

func TestPostgresDriverOwnsTheSQLSafetyTables(t *testing.T) {
	source, err := os.ReadFile("postgres.go")
	if err != nil {
		t.Fatal(err)
	}
	text := string(source)
	for _, required := range []string{"safeAdmit", "safeComplete", "insertOutboxSQL"} {
		if !strings.Contains(text, required) {
			t.Fatalf("postgres.go missing %s", required)
		}
	}
}

func TestReportRecordsWhichSideOwnsSafety(t *testing.T) {
	empty := newOpStats()
	riff := loadReport{
		backendID:         riffdbBackendID,
		clients:           1,
		profile:           "interactive",
		measuredElapsedNs: 1_000_000_000,
		seedNs:            1,
		aggregate:         empty,
		byOp:              map[string]*opStats{},
		workerCompleted:   []int{0},
	}.json()
	postgres := loadReport{
		backendID:         postgresBackendID,
		clients:           1,
		profile:           "interactive",
		measuredElapsedNs: 1_000_000_000,
		seedNs:            1,
		aggregate:         empty,
		byOp:              map[string]*opStats{},
		workerCompleted:   []int{0},
	}.json()
	if riff["safety_owner"] != "riffdbd_rust" {
		t.Fatalf("riff safety=%v", riff["safety_owner"])
	}
	if postgres["safety_owner"] != "golang_sql" {
		t.Fatalf("postgres safety=%v", postgres["safety_owner"])
	}
	if riff["evidentiary"] != false {
		t.Fatal("golang reports must not be evidentiary")
	}
	notes, _ := riff["notes"].([]string)
	joined := strings.Join(notes, " ")
	if !strings.Contains(joined, "No authorization, idempotency, audit, event, or outbox code runs in Go") {
		t.Fatalf("notes=%v", notes)
	}
}

func TestPercentilesTrackInjectedLatencies(t *testing.T) {
	histogram := newLatencyHistogram()
	for i := 0; i < 90; i++ {
		histogram.recordNs(100_000)
	}
	for i := 0; i < 9; i++ {
		histogram.recordNs(1_000_000)
	}
	histogram.recordNs(10_000_000)
	if histogram.percentileNs(50) < 100_000 || histogram.percentileNs(50) >= 1_000_000 {
		t.Fatalf("p50=%d", histogram.percentileNs(50))
	}
	if histogram.percentileNs(99) < 1_000_000 {
		t.Fatalf("p99=%d", histogram.percentileNs(99))
	}
}

File: examples/app-baseline/typescript/src/ids.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import test from "node:test";
import { fileURLToPath } from "node:url";

import { formatUuid, uuidFromOrdinal } from "./ids.js";
import { INTERACTIVE_WEIGHTS, LoadReport, OpStats } from "./load.js";
import { OBLIGATIONS, SCHEMA_SQL, safeInputFingerprint } from "./schema.js";
import { boardDenseOpenCount, fullScale, generateSeed, smokeScale } from "./seed.js";

const repoRoot = fileURLToPath(new URL("../../../../", import.meta.url));

test("uuid_from_ordinal matches rust layout", () => {
  const value = uuidFromOrdinal(0x10, 0);
  assert.equal(value.length, 16);
  assert.equal(value[0], 0x10);
  assert.equal(value[6], 0x70);
  assert.equal(value[8]! & 0xc0, 0x80);
  assert.equal(formatUuid(value), "10101010-1010-7010-8000-000000000000");
});

test("fingerprint is canonical and length prefixed", () => {
  const first = safeInputFingerprint([Buffer.from("a/b"), Buffer.from("c")]);
  const second = safeInputFingerprint([Buffer.from("a"), Buffer.from("b/c")]);
  const repeated = safeInputFingerprint([Buffer.from("a/b"), Buffer.from("c")]);
  assert.equal(first.length, 64);
  assert.equal(first, repeated);
  assert.notEqual(first, second);
  assert.equal(first.includes("a/b"), false);
});

test("smoke and full seed shapes", () => {
  const smoke = generateSeed(smokeScale());
  assert.equal(smoke.organizations.length, 2);
  assert.equal(smoke.scale.boardDenseOpen, 0);
  assert.ok(boardDenseOpenCount(smoke) < 50);
  const full = generateSeed(fullScale());
  assert.equal(boardDenseOpenCount(full), 600);
});

test("schema contains safe-app obligations", () => {
  for (const table of [
    "app_permission",
    "app_idempotency",
    "app_audit",
    "app_domain_event",
    "app_outbox_intent",
  ]) {
    assert.ok(SCHEMA_SQL.includes(`CREATE TABLE ${table}`));
  }
  assert.deepEqual(OBLIGATIONS, [
    "symbolic_operation_authorization",
    "idempotency_admission_and_equal_input_replay",
    "domain_mutation",
    "audit_and_provenance",
    "domain_event",
    "outbox_intent",
    "one_atomic_transaction",
  ]);
});

test("python schema tables match this schema", () => {
  const rust = readFileSync(`${repoRoot}examples/app-baseline/postgres/src/lib.rs`, "utf8");
  const rustTables = new Set([...rust.matchAll(/CREATE TABLE (\w+)/g)].map((match) => match[1]));
  const typescriptTables = new Set(
    [...SCHEMA_SQL.matchAll(/CREATE TABLE (\w+)/g)].map((match) => match[1]),
  );
  assert.deepEqual(typescriptTables, rustTables);
});

test("interactive weights are mostly reads", () => {
  const total = INTERACTIVE_WEIGHTS.reduce((sum, [, weight]) => sum + weight, 0);
  const writes = INTERACTIVE_WEIGHTS.filter(([name]) =>
    ["create_comment", "close_ticket_with_comment", "open_ticket_with_labels"].includes(name),
  ).reduce((sum, [, weight]) => sum + weight, 0);
  assert.ok(total - writes > writes);
  assert.equal(
    INTERACTIVE_WEIGHTS.some(([name]) => name === "swap_member_roles"),
    false,
  );
});

test("riffdb driver contains no typescript safety implementation", () => {
  const source = readFileSync(new URL("../src/riffdb.ts", import.meta.url), "utf8");
  for (const forbidden of [
    "app_permission",
    "app_idempotency",
    "app_audit",
    "app_domain_event",
    "app_outbox_intent",
    "PERMISSION_SQL",
    "IDEMPOTENCY_LOCK_SQL",
    "INSERT_AUDIT_SQL",
    "INSERT_EVENT_SQL",
    "INSERT_OUTBOX_SQL",
    "safeAdmit",
    "safeComplete",
    "safeInputFingerprint",
  ]) {
    assert.equal(source.includes(forbidden), false, forbidden);
  }
  assert.ok(source.includes("No TypeScript safety code"));
  assert.ok(source.includes("generated TicketDesk"));
});

test("postgres driver owns the sql safety tables", () => {
  const source = readFileSync(new URL("../src/postgres.ts", import.meta.url), "utf8");
  assert.ok(source.includes("safeAdmit"));
  assert.ok(source.includes("safeComplete"));
  assert.ok(source.includes("INSERT_OUTBOX_SQL"));
});

test("generated ticketdesk client is present", () => {
  const text = readFileSync(`${repoRoot}clients/typescript/ticketdesk/client.ts`, "utf8");
  assert.ok(text.includes("export class TicketDeskClient"));
  assert.ok(text.includes("createComment"));
});

test("report records which side owns safety", () => {
  const riff = new LoadReport(
    "riffdb_public_grpc",
    1,
    "interactive",
    1_000_000_000,
    1,
    new OpStats(),
    {},
    [0],
  ).json();
  const postgres = new LoadReport(
    "postgres_safe_app",
    1,
    "interactive",
    1_000_000_000,
    1,
    new OpStats(),
    {},
    [0],
  ).json();
  assert.equal(riff.safety_owner, "riffdbd_rust");
  assert.equal(postgres.safety_owner, "typescript_sql");
  assert.equal(riff.evidentiary, false);
  assert.ok(
    (riff.notes as string[]).join(" ").includes(
      "No authorization, idempotency, audit, event, or outbox code runs in TypeScript",
    ),
  );
});
